GroovyJarFile cannot resolve a symlink location in the classpath

### Current Behavior

Our build system uses a home grown plugin to inject alternative dependency coordinates for popular libraries managed internally and not aligned with the common public dependency coordinates. It is interesting to note that this system makes heavy use of symlinks in the process and the concrete target files provided retain the expected file names.

One case that is particularly problematic is the Groovy runtime. This is resolved by the GroovyJarFile construct in Gradle and it assumes a File reference is representing the actual JAR file to pattern match for the expected artifact name. The symlink name fails the pattern matching and causes the problems.

### Current Behavior

The error we see when this resolution fails is this:

```
Caused by: org.gradle.api.GradleException: Cannot infer Groovy class path because no Groovy Jar was found on class path:
 at org.gradle.api.tasks.GroovyRuntime$1.inferGroovyClasspath(GroovyRuntime.java:118)
        at org.gradle.api.tasks.GroovyRuntime$1.createDelegate(GroovyRuntime.java:108)
```


### Expected Behavior

The alternative classpath entry does not cause a failure and can be used in place of the public accepted coordinates.

### Context (optional)

We are failing to make use of certain components that rely on Groovy Runtime.

### Potential fix for the issue

The issue can be resolved if [the code found here](https://github.com/gradle/gradle/blob/master/platforms/jvm/plugins-groovy/src/main/java/org/gradle/api/internal/plugins/GroovyJarFile.java#L62-L66) performs a resolution of the real path instead of using the provided File object in raw form.

I performed an experiment adding this code to find that the issue is resolved.

```diff
     @Nullable
     public static GroovyJarFile parse(File file) {
+        try {
+            // Resolve a symlink file to the real location
+            file = file.toPath().toRealPath().toFile();
+        } catch (IOException e) {
+            // Let the code use the original File otherwise
+        }
         Matcher matcher = FILE_NAME_PATTERN.matcher(file.getName());
         return matcher.matches() ? new GroovyJarFile(file, matcher) : null;
     }
```


### Steps to Reproduce

Here is a very small github project that can demonstrate the issue:

https://github.com/Alex-Vol-Amz/groovy-jar-symlink-issue


### Gradle version

7.6.1

### Build scan URL (optional)

_No response_

### Your Environment (optional)

Currently experiencing this with Gradle 7 and plan to continue using with Gradle 8. Hopefully a fix can be provided in the 8.x time frame for new releases.

--- a/platforms/jvm/plugins-groovy/src/main/java/org/gradle/api/internal/plugins/GroovyJarFile.java
+++ b/platforms/jvm/plugins-groovy/src/main/java/org/gradle/api/internal/plugins/GroovyJarFile.java
@@ -19,6 +19,7 @@ import org.gradle.util.internal.VersionNumber;
 
 import javax.annotation.Nullable;
 import java.io.File;
+import java.io.IOException;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -61,6 +62,14 @@ public class GroovyJarFile {
 
     @Nullable
     public static GroovyJarFile parse(File file) {
+        try {
+            if (file.getName().contains("groovy")) {
+                // Resolve a symlink file to the real location
+                file = file.toPath().toRealPath().toFile();
+            }
+        } catch (IOException e) {
+            // Let the code use the original File otherwise
+        }
         Matcher matcher = FILE_NAME_PATTERN.matcher(file.getName());
         return matcher.matches() ? new GroovyJarFile(file, matcher) : null;
     }

--- a/platforms/jvm/plugins-groovy/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
+++ b/platforms/jvm/plugins-groovy/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
@@ -15,10 +15,15 @@
  */
 package org.gradle.api.internal.plugins
 
+import java.nio.file.Files
 import org.gradle.util.internal.VersionNumber
 import spock.lang.Specification
+import spock.lang.TempDir
 
 class GroovyJarFileTest extends Specification {
+    @TempDir
+    def tempDir
+
     def "parse non-Groovy file"() {
         expect:
         GroovyJarFile.parse(new File("groovy-other-2.0.5.jar")) == null
@@ -85,5 +90,33 @@ class GroovyJarFileTest extends Specification {
         "/lib/groovy-5.0.2-indy.jar" | "5.0.2" | "org.apache.groovy:groovy:5.0.2:indy"
     }
 
+    def "parse symlinked Jar"(String symlinkName, String fileName, String version, String dependencyNotation) {
+        File symlinkFile = new File(tempDir.toFile(), symlinkName)
+        File actualFile = new File(tempDir.toRealPath().toFile(), fileName)
+
+        if (!actualFile.exists()) {
+            actualFile.createNewFile()
+        }
+        if (!symlinkFile.exists()) {
+            Files.createSymbolicLink(symlinkFile.toPath(), actualFile.toPath())
+        }
+
+        def jar = GroovyJarFile.parse(symlinkFile)
+
+        expect:
+        jar != null
+        jar.file == actualFile
+        jar.baseName == "groovy"
+        jar.version == VersionNumber.parse("$version")
+        !jar.groovyAll
+        !jar.indy
+        jar.dependencyNotation == "$dependencyNotation"
+
+        where:
+        symlinkName            | fileName           | version | dependencyNotation
+        "Maven-groovy-2.x.jar" | "groovy-2.0.5.jar" | "2.0.5" | "org.codehaus.groovy:groovy:2.0.5"
+        "Maven-groovy-4.x.jar" | "groovy-4.1.0.jar" | "4.1.0" | "org.apache.groovy:groovy:4.1.0"
+        "Maven-groovy-5.x.jar" | "groovy-5.0.2.jar" | "5.0.2" | "org.apache.groovy:groovy:5.0.2"
+    }
 
 }

--- a/platforms/jvm/plugins-groovy/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
+++ b/platforms/jvm/plugins-groovy/src/test/groovy/org/gradle/api/internal/plugins/GroovyJarFileTest.groovy
@@ -15,14 +15,14 @@
  */
 package org.gradle.api.internal.plugins
 
-import java.nio.file.Files
+import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
 import org.gradle.util.internal.VersionNumber
+import org.junit.Rule
 import spock.lang.Specification
-import spock.lang.TempDir
 
 class GroovyJarFileTest extends Specification {
-    @TempDir
-    def tempDir
+    @Rule
+    TestNameTestDirectoryProvider tmpDir = new TestNameTestDirectoryProvider(getClass())
 
     def "parse non-Groovy file"() {
         expect:
@@ -91,15 +91,8 @@ class GroovyJarFileTest extends Specification {
     }
 
     def "parse symlinked Jar"(String symlinkName, String fileName, String version, String dependencyNotation) {
-        File symlinkFile = new File(tempDir.toFile(), symlinkName)
-        File actualFile = new File(tempDir.toRealPath().toFile(), fileName)
-
-        if (!actualFile.exists()) {
-            actualFile.createNewFile()
-        }
-        if (!symlinkFile.exists()) {
-            Files.createSymbolicLink(symlinkFile.toPath(), actualFile.toPath())
-        }
+        def actualFile = tmpDir.file(fileName).touch()
+        def symlinkFile = tmpDir.file(symlinkName).createLink(actualFile)
 
         def jar = GroovyJarFile.parse(symlinkFile)
 

